ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -82,21 +82,19 @@ data InProgressState = InProgressState
 newInProgress :: IO InProgressState
 newInProgress = InProgressState <$> newTVarIO 0 <*> newTVarIO 0 <*> STM.newIO
 
-recordProgress :: InProgressState -> NormalizedFilePath -> (Int -> Int) -> STM ()
+recordProgress :: InProgressState -> NormalizedFilePath -> (Int -> Int) -> IO ()
 recordProgress InProgressState{..} file shift = do
-    done <- readTVar doneVar
-    todo <- readTVar todoVar
-    (prev, new) <- STM.focus alterPrevAndNew file currentVar
-    let (done',todo') =
-            case (prev,new) of
-                (Nothing,0) -> (done+1, todo+1)
-                (Nothing,_) -> (done,   todo+1)
-                (Just 0, 0) -> (done  , todo)
-                (Just 0, _) -> (done-1, todo)
-                (Just _, 0) -> (done+1, todo)
-                (Just _, _) -> (done  , todo)
-    writeTVar todoVar todo'
-    writeTVar doneVar done'
+    (prev, new) <- atomicallyNamed "recordProgress" $ STM.focus alterPrevAndNew file currentVar
+    atomicallyNamed "recordProgress2" $ do
+        done <- readTVar doneVar
+        todo <- readTVar todoVar
+        case (prev,new) of
+            (Nothing,0) -> writeTVar doneVar (done+1) >> writeTVar todoVar (todo+1)
+            (Nothing,_) -> writeTVar todoVar (todo+1)
+            (Just 0, 0) -> pure ()
+            (Just 0, _) -> writeTVar doneVar (done-1)
+            (Just _, 0) -> writeTVar doneVar (done+1)
+            (Just _, _) -> pure()
```

Comment:
```suggestion
    atomicallyNamed "recordProgress2" $ do
        case (prev,new) of
            (Nothing,0) -> modifyTVar doneVar (+1) >> modifyTVar todoVar (+1)
            (Nothing,_) -> modifyTVar todoVar (+1)
            (Just 0, 0) -> pure ()
            (Just 0, _) -> modifyTVar doneVar (-1)
            (Just _, 0) -> modifyTVar doneVar (+1)
            (Just _, _) -> pure()
```

This seems like it should be even better, since it avoids reading the variables unless it writes to them, and I _think_ that STM will retry a transaction even if it only _reads_ a changed variable?

Reply:
Neat, and definitely better. Thanks!
```
STM transaction statistics (2021-12-12 22:11:20.016977 UTC):
Transaction               Commits    Retries      Ratio
_anonymous_                 15227        116       0.01
action queue - pop              2          2       1.00
actionQueue - done              2          0       0.00
actionQueue - peek             29          0       0.00
actionQueue - push              2          0       0.00
builder                    282373        771       0.00
compute                     16882         32       0.00
debouncer                    6864        215       0.03
define - dirtyKeys          16900          0       0.00
define - read 1             10710          3       0.00
define - read 2              6254          3       0.00
define - write               6248          1       0.00
diagnostics - hidden         6893         10       0.00
diagnostics - publish        4006        200       0.05
diagnostics - read           6901          1       0.00
diagnostics - update         6893         22       0.00
incDatabase                 10966          0       0.00
lastValueIO 4                2200          0       0.00
lastValueIO 5                2200          0       0.00
recordProgress              31238        387       0.01
recordProgress2             31238         79       0.00
updateReverseDeps           64994        387       0.01
```